File: CPP/SerializeAndDeserializeBinaryTree.cpp
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
#include <vector>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

class Codec {
public:
  // Level-order encoding: node value, or '#' for a missing child.
  std::string serialize(TreeNode *root) {
    if (!root)
      return "";
    std::ostringstream out;
    std::queue<TreeNode *> q;
    q.push(root);
    bool first = true;
    while (!q.empty()) {
      TreeNode *node = q.front();
      q.pop();
      if (!first)
        out << ",";
      first = false;
      if (node) {
        out << node->val;
        q.push(node->left);
        q.push(node->right);
      } else {
        out << "#";
      }
    }
    return out.str();
  }

  TreeNode *deserialize(std::string data) {
    if (data.empty())
      return nullptr;
    std::vector<std::string> tokens;
    std::stringstream ss(data);
    std::string tok;
    while (std::getline(ss, tok, ','))
      tokens.push_back(tok);
    TreeNode *root = new TreeNode(std::stoi(tokens[0]));
    std::queue<TreeNode *> q;
    q.push(root);
    size_t i = 1;
    while (!q.empty()) {
      TreeNode *node = q.front();
      q.pop();
      if (tokens[i] != "#") {
        node->left = new TreeNode(std::stoi(tokens[i]));
        q.push(node->left);
      }
      ++i;
      if (tokens[i] != "#") {
        node->right = new TreeNode(std::stoi(tokens[i]));
        q.push(node->right);
      }
      ++i;
    }
    return root;
  }
};

std::string toList(TreeNode *root) {
  std::vector<std::string> out;
  std::queue<TreeNode *> q;
  q.push(root);
  while (!q.empty()) {
    TreeNode *node = q.front();
    q.pop();
    if (node) {
      out.push_back(std::to_string(node->val));
      q.push(node->left);
      q.push(node->right);
    } else {
      out.push_back("null");
    }
  }
  while (!out.empty() && out.back() == "null")
    out.pop_back();
  std::string s = "[";
  for (size_t i = 0; i < out.size(); ++i) {
    if (i)
      s += ", ";
    s += out[i];
  }
  return s + "]";
}

int main() {
  Codec c;
  TreeNode *root = new TreeNode(1);
  root->left = new TreeNode(2);
  root->right = new TreeNode(3);
  root->right->left = new TreeNode(4);
  root->right->right = new TreeNode(5);
  std::string data = c.serialize(root);
  std::cout << data << "\n";
  std::cout << toList(c.deserialize(data))
            << "\n"; // [1, 2, 3, null, null, 4, 5]
  std::cout << "\"" << c.serialize(c.deserialize(c.serialize(nullptr)))
            << "\"\n"; // ""
  std::cout << toList(c.deserialize(c.serialize(new TreeNode(7))))
            << "\n"; // [7]
  return 0;
}

File: Python/SerializeAndDeserializeBinaryTree.py
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Codec:
    # Level-order (BFS) encoding: each node emits its value, each missing child emits '#'.
    # Decoding replays the queue, attaching children as their tokens are consumed.
    def serialize(self, root: Optional[TreeNode]) -> str:
        if not root:
            return ""
        out = []
        q = deque([root])
        while q:
            node = q.popleft()
            if node:
                out.append(str(node.val))
                q.append(node.left)
                q.append(node.right)
            else:
                out.append("#")
        return ",".join(out)

    def deserialize(self, data: str) -> Optional[TreeNode]:
        if not data:
            return None
        tokens = data.split(",")
        root = TreeNode(int(tokens[0]))
        q = deque([root])
        i = 1
        while q:
            node = q.popleft()
            if tokens[i] != "#":
                node.left = TreeNode(int(tokens[i]))
                q.append(node.left)
            i += 1
            if tokens[i] != "#":
                node.right = TreeNode(int(tokens[i]))
                q.append(node.right)
            i += 1
        return root


def to_list(root):
    out = []
    q = deque([root])
    while q:
        node = q.popleft()
        out.append(node.val if node else None)
        if node:
            q.append(node.left)
            q.append(node.right)
    while out and out[-1] is None:
        out.pop()
    return out


if __name__ == "__main__":
    c = Codec()
    # [1,2,3,null,null,4,5]
    root = TreeNode(1, TreeNode(2), TreeNode(3, TreeNode(4), TreeNode(5)))
    data = c.serialize(root)
    print(data)
    print(to_list(c.deserialize(data)))  # [1, 2, 3, None, None, 4, 5]
    print(repr(c.serialize(c.deserialize(c.serialize(None)))))  # ''
    single = TreeNode(7)
    print(to_list(c.deserialize(c.serialize(single))))  # [7]
